lib/tasks/eu_exit_business_finder.rake

```diff
@@ -0,0 +1,106 @@
+namespace :eu_exit_business_finder do
+  task update_subscriber_list_titles_and_facet_values: [:environment] do
+    subscriber_lists = SubscriberList
+      .where("links::text != '{}'")
+      .where("ARRAY(SELECT json_object_keys(links)) <@ Array[:keys]", keys: [:facet_values])
+    update_facet_values(subscriber_lists)
+  end
+
+  # Iterate through the affected subscriber lists and if we encounter any
+  # facet values that need replacing with a newer one, we insert the new one
+  # to the array (inserted 'in place' to retain original order). Then remove
+  # the old facet values in bulk and write only the unique values that remain.
+  def update_facet_values(subscriber_lists)
+    subscriber_lists.each do |list|
+      links = list.links[:facet_values][:any]
+      updated_links = links.dup # copy the links to avoid corrupting the loop
+      links.each_with_index do |value, index|
+        replacement = EuExitFacetMigrationConfig::facet_values_to_replace.find { |replace_val| replace_val[:old_facet_value] == value }
+        updated_links.insert(index, replacement[:new_facet_value]) if replacement
+      end
+      updated_links = updated_links - EuExitFacetMigrationConfig::facet_values_to_remove
+      updated_links = updated_links.uniq
+      list.write_attribute('links', facet_values: { any: updated_links })
+      list.write_attribute('title', title_from_facets(updated_links))
+      list.save
+    end
+  end
+
+  # Title is constructed from finder content item, so we have to grab the JSON.
+  def title_from_facets(links)
+    json_response = JSON.parse(Net::HTTP.get(URI('https://www.gov.uk/api/content/find-eu-exit-guidance-business')))
```

Comment:
Would it be possible to use `gds-api-adapters` to get this from publishing-api?  It would prevent any issues with egress rules on machines in production.

Reply:
Good shout. 👍 Shall I use [publishing_api_v2](https://github.com/alphagov/gds-api-adapters/blob/master/lib/gds_api/publishing_api_v2.rb)?

Reply:
Yes, that should have a suitable method to get content

Reply:
Any reason why publishing API and not content API? I thought the former was for writing and the latter was for retrieving...?

Reply:
Content API is a service that's available to the public for accessing data.  As far as I'm aware, it's not actually used by any of our internal services.

Reply:
Our docs seem to suggest they do (if front-end apps count as internal services):

> This API is used by front-end apps but is also exposed externally at /api/content/<path>, such as https://www.gov.uk/api/content/take-pet-abroad

https://docs.publishing.service.gov.uk/apis/content-store.html#reading-and-writing-to-the-store

Also could you explain "It would prevent any issues with egress rules on machines in production"? I don't know what that means :)

Reply:
>Also could you explain "It would prevent any issues with egress rules on machines in production"? I don't know what that means :)
I believe some machines in AWS have rules about what ports and IP addresses access into (ingress) and access out of (egress) machines is allowed, for security reasons.  It may be that the production machines can't make a HTTP request to an external site (I'm not certain).

Reply:
I've just learned two words today - thanks for that!

Not exactly sure what I'm doing here - replacing the line with: 

```ruby
json_response = JSON.parse(GdsApi::ContentStore.new(Plek.new.find("content-store")).content_item('/find-eu-exit-guidance-business'))
```

...makes a call to `http://content-store.dev.gov.uk/content/find-eu-exit-guidance-business`. Is that what we want? I'm worried I won't catch any issues encountered if I'm just stubbing the network request to return what I want 🤔

Reply:
That would work.  If it were me, I'd use the publishing API, since you wouldn't then need to parse the JSON:
```
GdsApi::PublishingApiV2.new(Plek.new.find('publishing-api')).get_content('content_id_of_finder')
```
But I guess it doesn't really matter too much either way.

Even better, I'd probably define a `publishing_api` method, then just use that (it could even be defined in `./app/lib/services.rb`, so you can then access it in the tests using `Services.publishing_api`):
```
def publishing_api
   GdsApi::PublishingApiV2.new(Plek.new.find('publishing-api'))
end

....

publishing_api.get_content('content_id_of_finder')
```

I'd strongly suggest stubbing the request (hence the need for the fixture file) otherwise the tests will start failing if the live content item is changed elsewhere.  The test should be checking the method performs the desired functionality, which it can do with mock data and doesn't necessarily need real data.  This gives the advantage that the tests will still pass if you were to change the values in `facet_values_to_replace` at some point in the future but leave the code unchanged (my belief is this rake task should stick around until the finder is removed, as I'm sure the sectors will get changed again at some point 🤔).

Reply:
Oh I'm already stubbing, it's just I'm stubbing the https://www.gov.uk/api/content/find-eu-exit-guidance-business endpoint (which I can manually check and make sure it will return the right thing in production) rather than stubbing `http://content-store.dev.gov.uk/content/find-eu-exit-guidance-business` (which I can't manually check, so stubbing locally is fine but I have less confidence it will work on production).

Good tip about using publishing API to avoid having to parse the JSON. Also interesting you say the code will stick around, as I'd imagined us deleting it shortly after use, but agree it makes sense to keep for as long as the finder exists.

Reply:
Have now pushed up an update that should resolve this conversation 👍